### ISSUE TITLE
NoSuchMethodError on server start-up

### ISSUE THREAD

**Issue by SlimeDog** [PlaceholderAPI/PlaceholderAPI#376]
[Request change (Wiki)]: https://github.com/PlaceholderAPI/PlaceholderAPI/issues/new?template=change_request_wiki.md
[Request change (PlaceholderAPI)]: https://github.com/PlaceholderAPI/PlaceholderAPI/issues/new?template=change_request_placeholderapi.md
[Spigot]: https://www.spigotmc.org/resources/6245/
[Jenkins]: http://ci.extendedclip.com/job/PlaceholderAPI/
[issues]: https://github.com/PlaceholderAPI/PlaceholderAPI/issues

## Please read
This template is only for reporting bugs of PlaceholderAPI!  
If you want to request changes to the code of PlaceholderAPI, use the [Request change (PlaceholderAPI)] template.  
For changes to the wiki use the [Request change (Wiki)] template.

Please also make sure that you use the [latest release][Spigot] or the latest [developement build][Jenkins] and that your bug isn't already reported on the [issues] page.

**Please report issues with expansions/placeholders to the corresponding issue tracker! Only report it here if you can't find any place to report it!**

### Issue
> What is the issue? Describe it like you would tell a friend.

PAPI generates a `NoSuchMethodError` message on start-up.

### Expected behaviour
> What should PlaceholderAPI do?

Successful expansion loading.

### Actual behaviour
> What does PlaceholderAPI actually do?

On server start-up
```
[10:53:41] [Server thread/INFO]: [PlaceholderAPI] Loading PlaceholderAPI v2.10.7-DEV-29
[10:53:48] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: aac
[10:53:48] [Server thread/INFO]: [AAC] Hooked into PlaceholderAPI
[10:54:05] [Server thread/INFO]: [PlaceholderAPI] Enabling PlaceholderAPI v2.10.7-DEV-29
[10:54:06] [Server thread/INFO]: [StatisticEditor] PlaceholderAPI found, registering hook
[10:54:06] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: stat
[10:54:07] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: spark
[10:54:07] [Server thread/INFO]: [spark] Registered PlaceholderAPI placeholders
[10:54:07] [Server thread/INFO]: [HelpGUI] Checking for PlaceholderAPI...
[10:54:07] [Server thread/INFO]: [HelpGUI] Hooked into PlaceholderAPI.
[10:54:09] [Server thread/INFO]: [DeluxeMenus] Hooked into PlaceholderAPI!
[10:54:09] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: DeluxeMenus
[10:54:39] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: insights
[10:54:39] [Server thread/INFO]: Successfully hooked into PlaceholderAPI!
[10:54:40] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: frp
[10:54:42] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: residence
[10:54:42] [Server thread/INFO]: ^[[0;32;1m[^[[0;33;22mResidence^[[0;32;1m]^[[0;37;22m PlaceholderAPI hooked.^[[m
[10:54:42] [Server thread/INFO]: ^[[0;32;1m[^[[0;33;22mResidence^[[0;32;1m]^[[0;37;22m PlaceholderAPI was found - Enabling capabilities.^[[m
[10:54:42] [Server thread/INFO]: [ItemJoin] Hooked into { Multiverse-Core, Multiverse-Inventories, WorldGuard, PlaceholderAPI, Vault }
[10:54:42] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: biomeremap
[10:54:43] [Server thread/INFO]: [Server Tutorial Plus] PlaceholderAPI has been found!
[10:54:44] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: denizen
[10:54:44] [Server thread/INFO]: [CombatLogX] Enabling expansion 'PlaceholderAPI Compatibility v15.0'...
[10:54:44] [Server thread/INFO]: Successfully hooked into PlaceholderAPI v2.10.7-DEV-29
[10:54:44] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: combatlogx
[10:54:44] [Server thread/INFO]: Successfully hooked into PlaceholderAPI v2.10.7-DEV-29
[10:54:48] [Server thread/INFO]: ^[[0;33;22m[MobHunting] ^[[mEnabling compatibility with PlaceholderAPI (2.10.7-DEV-29).^[[m
[10:54:48] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: mobhunting
[10:54:50] [Server thread/INFO]: [PlaceholderAPI] Placeholder expansion registration initializing...
[10:54:50] [Server thread/INFO]: [BentoBox] Hooking with PlaceholderAPI...
[10:54:50] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: bentobox
[10:54:50] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: level
[10:54:50] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: bskyblock
[10:54:50] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: challenges
[10:54:57] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: plan
[10:54:59] [Server thread/WARN]: [PlaceholderAPI] Loaded class com.earth2me.essentials.Essentials from Essentials v2.18.0.7 which is not a depend, softdepend or
 loadbefore of this plugin.
[10:54:59] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: vanish
[10:54:59] [Server thread/WARN]: [PlaceholderAPI] Loaded class com.onarandombox.MultiverseCore.MultiverseCore from Multiverse-Core v4.1.1-SNAPSHOT-b785 which is
 not a depend, softdepend or loadbefore of this plugin.
[10:54:59] [Server thread/INFO]: [PlaceholderAPI] Successfully registered expansion: multiverse
[10:54:59] [Server thread/WARN]: [PlaceholderAPI] Loaded class com.hmmcrunchy.thirst.Thirst from Thirst v2.11.1 which is not a depend, softdepend or loadbefore
of this plugin.
[10:54:59] [Server thread/WARN]: [PlaceholderAPI] Task #7514 for PlaceholderAPI v2.10.7-DEV-29 generated an exception
java.lang.NoSuchMethodError: me.clip.placeholderapi.PlaceholderAPI.registerPlaceholderHook(Ljava/lang/String;Lme/clip/placeholderapi/PlaceholderHook;)Z
        at com.extendedclip.papi.expansion.thirst.ThirstExpansion.register(ThirstExpansion.java:29) ~[?:?]
        at me.clip.placeholderapi.expansion.manager.LocalExpansionManager.register(LocalExpansionManager.java:233) ~[?:?]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:257) ~[?:?]
        at java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:248) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.count(ReferencePipeline.java:605) ~[?:?]
        at me.clip.placeholderapi.expansion.manager.LocalExpansionManager.lambda$null$2(LocalExpansionManager.java:291) ~[?:?]
        at org.bukkit.craftbukkit.v1_16_R1.scheduler.CraftTask.run(CraftTask.java:81) ~[spigot-1.16.1.jar-2020-07-24-1727:git-Spigot-9639cf7-7c03d25]
        at org.bukkit.craftbukkit.v1_16_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:400) ~[spigot-1.16.1.jar-2020-07-24-1727:git-Spigot-9639cf7-7c03d25]
        at net.minecraft.server.v1_16_R1.MinecraftServer.b(MinecraftServer.java:1061) ~[spigot-1.16.1.jar-2020-07-24-1727:git-Spigot-9639cf7-7c03d25]
        at net.minecraft.server.v1_16_R1.DedicatedServer.b(DedicatedServer.java:354) ~[spigot-1.16.1.jar-2020-07-24-1727:git-Spigot-9639cf7-7c03d25]
        at net.minecraft.server.v1_16_R1.MinecraftServer.a(MinecraftServer.java:1009) ~[spigot-1.16.1.jar-2020-07-24-1727:git-Spigot-9639cf7-7c03d25]
        at net.minecraft.server.v1_16_R1.MinecraftServer.v(MinecraftServer.java:848) ~[spigot-1.16.1.jar-2020-07-24-1727:git-Spigot-9639cf7-7c03d25]
        at net.minecraft.server.v1_16_R1.MinecraftServer.lambda$0(MinecraftServer.java:164) ~[spigot-1.16.1.jar-2020-07-24-1727:git-Spigot-9639cf7-7c03d25]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```
Secondarily, this should probably be ERROR rather than WARN, since the result is failure.

### How to reproduce
> What steps did you made, to get this bug?

1. Update from PAPI dev build 22 to dev build 29 
2. Start server

The same sequence works correctly with PAPI dev build 22

**Comment by SlimeDog** [comment]
Issue still exists in PAPI build 30. Is this a problem with the Thirst expansion or PAPI itself?

**Comment by SlimeDog** [comment]
A little more research and isolation. Similar errors are produced by any of the following:
- DiseasePlaceholders-0.0.1.jar
- Expansion-Skript-1.0.0.jar
- Expansion-Thirst-1.1.0.jar
- PAPI-CoordinateTool-1.1.0.jar

Without them, my other 10 expansions load. These are installed
```
Expansion-LuckPerms-5.0-R2.jar
Expansion-Multiverse-1.0.1.jar
Expansion-Quests-1.1.jar
Expansion-Statistics-1.7.0.jar
Expansion-ViaVersion.jar
PAPI-Expansion-Essentials-1.2.4.jar
PAPI-Expansion-Player-1.7.2.jar
PAPI-Expansion-Server-2.2.3.jar
PAPI-Expansion-Statistic-1.9.2.jar
PAPI-Expansion-Vault-1.6.0.jar
VanishPlaceholders.jar
WorldGuard-Expansion-1.4.2.jar
```
However, `papi ecloud list installed` still produces nothing.
```
>papi list
[19:06:37] [Server thread/INFO]: 27 placeholder hook(s) active: 
  DeluxeMenus, aac, bentobox, biomeremap, bskyblock, challenges, combatlogx, denizen, essentials, frp
  insights, level, luckperms, mobhunting, multiverse, plan, player, quests, residence, server
  spark, stat, statistic, vanish, vault, viaversion, worldguard
>papi ecloud enable
[19:07:06] [Server thread/INFO]: [PlaceholderAPI] Fetching available expansion information...
[19:07:06] [Server thread/INFO]: The eCloud Manager has been enabled
>papi ecloud list installed
>

**Comment by Sxtanna** [comment]
The issue of the expansions not loading is not an issue of PAPI, they will need to be updated.

However, ecloud list not functioning is a completely separate thing.

**Comment by SlimeDog** [comment]
> The issue of the expansions not loading is not an issue of PAPI, they will need to be updated

OK. I will open issues for them. No doubt there are others that I don't employ.

And a new issue for `papi ecloud list installed`

**Comment by Sxtanna** [comment]
Ok, I've identified the issue with the list command doing nothing, that will be updated in the next build

**Comment by Sxtanna** [comment]
> And a new issue for papi ecloud list installed

Oh ok, cool.

**Comment by SlimeDog** [comment]
Well, I won't open that one, since you've resolved it.

**Comment by Sxtanna** [comment]
No, please do. It's useful for logging.

**Comment by SlimeDog** [comment]
> No, please do. It's useful for logging.

Done. #377

**Comment by SlimeDog** [comment]
> I will open issues for them.

What do they need to do? Is it a deprecated method? Do they need to update to later PAPI? Please be specific as possible, and I will open tickets for the identified ones.

**Comment by PiggyPiglet** [comment]
> > I will open issues for them.
> 
> What do they need to do? Is it a deprecated method? Do they need to update to later PAPI? Please be specific as possible, and I will open tickets for the identified ones.

Any expansion that handles it's own registration, may not work. i.e. expansions that register themselves, via PlaceholderAPI#registerPlaceholderHook, will cease to function. In general, any expansion or plugin that uses the following methods:
- PlaceholderAPI#register*
- PlaceholderAPI#unregister*
will most likely break.

**Comment by SlimeDog** [comment]
Oh. So you are handling them?

For reference:
- When were those method deprecated?
- What method(s) replace them (ie., what methods need to be used)?

**Comment by PiggyPiglet** [comment]
> Oh. So you are handling them?
> 
> For reference:
> 
> * When were those method deprecated?
> * What method(s) replace them (ie., what methods need to be used)?

- They weren't deprecated, which is why I'm personally updating expansions.
- Registration is handled via papi itself, not the expansion, so there's no replacement.

**Comment by Sxtanna** [comment]
> Registration is handled via papi itself, not the expansion, so there's no replacement.

This is not entirely true.

If your expansion is compiled into your plugin, you should only ever have been calling 
`PlaceholderExpansion#register`

This is the *replacement* to those other methods. Even though its been there the whole time.. lol

**Comment by Sxtanna** [comment]
Also, this is technically closed since 07cd34412309253486dfb4d411c73695c783a829
